### PR TITLE
Update Systemd security settings

### PR DESCRIPTION
**Pull request by rahulsundaram**
This is a followup to https://src.fedoraproject.org/rpms/fwupd/pull-request/12.  I am proposing we add some additional security settings.  

https://www.freedesktop.org/software/systemd/man/latest/systemd.exec.html has detailed information on each of these settings including the version of Systemd where they were introduced so that you take into account RHEL and any other distros.

Please let me know if you want to drop any of these and I can adjust accordingly.

Type of pull request:
- [ ] Feature

### CODE REVIEW

File: data/fwupd.service.in [hunk 1]
```diff
@@ -12,8 +12,24 @@ RuntimeDirectory=@motd_dir@
...
+ProtectClock=yes
+ProtectControlGroups=yes
 ProtectHome=yes
+ProtectHostname=yes
+ProtectKernelLogs=yes
+ProtectKernelModules=yes
+ProtectKernelTunables=yes
+ProtectProc=invisible
 ProtectSystem=full
-SystemCallFilter=~@mount
```
Comment (superm1) [inline data/fwupd.service.in:18]: Why is this dropped?

Comment (rahulsundaram) [reply data/fwupd.service.in:18]: This was inadvertently dropped while resorting the list.  Fixed

**Review by superm1 — COMMENTED** [review]
I don't have any problems with any of these new protections; but I do have a problem that they're all being added directly into fwupd.service.in.  Our minimum version of systemd is 211 and a lot of these were introduced later.

We have a `@dynamic_options@` for this purpose which is populated by `data/meson.build`.

Actually some of these are already covered by that too!

So I think we need to bump our minimum systemd version and then clean some of that up.  @hughsie what do you think we can bump to?  249 OK?

**Comment by hughsie** [comment]
I think in master we can just depend on a super new systemd, I don't plan on backporting 2.0.x into Fedora 40 or RHEL 9 (because of the ABI bump too) for instance. It depends if we want the new systemd options in RHEL 9/Fedora 40.

I've done the bump to 249 to see what explodes in CI: https://github.com/fwupd/fwupd/pull/6863

**Comment by hughsie** [comment]
@rahulsundaram can you rebase against main, fix the conflicts and force push your branch please.

**Comment by rahulsundaram** [comment]
> @rahulsundaram can you rebase against main, fix the conflicts and force push your branch please.

Done.  Let me know if you want any other changes.  Thanks

**Comment by hughsie** [comment]
@rahulsundaram what happens if a key there isn't present in an older systemd version -- do we get a warning?

**Comment by rahulsundaram** [comment]
Yes, Older versions of Systemd will warn but not break the service outright if it sees a setting it doesn't grok

**Comment by hughsie** [comment]
> will warn but not break the service

Have you got an example about what a warning looks like? I'm worried that users are going to report bugs.

**Comment by superm1** [comment]
> @rahulsundaram what happens if a key there isn't present in an older systemd version -- do we get a warning?

I was hoping bumping to 249 as a min requirement we cover this. Are any of those after 249?

**Comment by rahulsundaram** [comment]
> > will warn but not break the service
> 
> Have you got an example about what a warning looks like? I'm worried that users are going to report bugs.

Yep.  This is an example from CentOS 7 running systemd-219-57.el7_5.3.x86_64.  This setting was introduced in version 247.

`systemd[1]: [/etc/systemd/system/unbound.service.d/override.conf:2] Unknown lvalue 'ProtectProc' in section 'Service'`

**Comment by hughsie** [comment]
@rahulsundaram if you change the fwupd.service to have `fwupd --vv --no-timestamp` and then run with the old and new settings is there any change on your machine? If not, I think we should merge these.

**Comment by rahulsundaram** [comment]
No change other than warnings along the lines of 

```
Feb 29 09:54:24 centos systemd[1]: /usr/lib/systemd/system/fwupd.service:28: Unknown lvalue 'ProtectClock' in section 'Service'
Feb 29 09:54:24 centos systemd[1]: /usr/lib/systemd/system/fwupd.service:31: Unknown lvalue 'ProtectHostname' in section 'Service'
Feb 29 09:54:24 centos systemd[1]: /usr/lib/systemd/system/fwupd.service:32: Unknown lvalue 'ProtectKernelLogs' in section 'Service'
Feb 29 09:54:24 centos systemd[1]: /usr/lib/systemd/system/fwupd.service:35: Unknown lvalue 'ProtectProc' in section 'Service'
```

**Comment by rahulsundaram** [comment]
> > No change other than warnings along the lines of
> 
> What about on a machine that _does_ support those options? I'm specifically worried about ProtectProc.

Good point.  With Verbose output, I did notice some differences and I have toggled a couple of settings.  Note that I had to use `-vv` instead of `--vv` from your suggested command.

**Comment by superm1** [comment]
> > > No change other than warnings along the lines of
> > 
> > 
> > What about on a machine that _does_ support those options? I'm specifically worried about ProtectProc.
> 
> Good point. With Verbose output, I did notice some differences and I have toggled a couple of settings. Note that I had to use `-vv` instead of `--vv` from your suggested command.

To be clear; not _all_ differences are bad.  We just have to make sure that they don't cause _other_ downstream impact.

File: data/fwupd.service.in [hunk 2]
```diff
@@ -12,11 +12,25 @@ RuntimeDirectory=@motd_dir@
...
+ProtectKernelTunables=yes
+ProtectProc=invisible
+ProtectSystem=full
+RestrictNamespaces=yes
 RestrictRealtime=yes
+RestrictSUIDSGID=yes
+SystemCallArchitectures=native
 SystemCallFilter=~@mount
 @dynamic_options@
+d
```
Comment (superm1) [inline data/fwupd.service.in:36]: spurious d

**Comment by superm1** [comment]
Can you please squash all your commits together as one?

**Review by superm1 — APPROVED** [review]

**Comment by rahulsundaram** [comment]
> To be clear; not _all_ differences are bad. We just have to make sure that they don't cause _other_ downstream impact.

Fair, I am just being conservative since I am not as familiar enough with this output to be aware of the exact downstream impact.  I am not looking to cover the most comprehensive settings I could find.  Just the common ones that should move the needle

**Comment by superm1** [comment]
> 
> > To be clear; not _all_ differences are bad. We just have to make sure that they don't cause _other_ downstream impact.
> 
> Fair, I am just being conservative since I am not as familiar enough with this output to be aware of the exact downstream impact.  I am not looking to cover the most comprehensive settings I could find.  Just the common ones that should move the needle

Thanks for the contribution!

**Comment by rahulsundaram** [comment]
Thank you both for your reviews.

@hughsie I would like this to available for Fedora 40/Rawhide.  Should I continue with the downstream PR or are you handling it?

**Comment by hughsie** [comment]
@superm1 what do you think about backporting these to 1.9.x?

**Comment by superm1** [comment]
I'm hesitant until we've really given more testing to them.

**Comment by hughsie** [comment]
I think I share that view too.

**Comment by rahulsundaram** [comment]
That's fair.  Do we have a timeline for the 2.0.0 release for my understanding? Just wondering if it is worth doing a downstream patch for Fedora 40 meanwhile or just let the upstream release roll through whenever.

**Comment by superm1** [comment]
> That's fair. Do we have a timeline for the 2.0.0 release for my understanding? Just wondering if it is worth doing a downstream patch for Fedora 40 meanwhile or just let the upstream release roll through whenever.

I don't see a world that something is safer in Fedora than it is in the upstream release.  I think the same risk applies.  Given Fedora 40 is around the corner, I don't think we want to potentially break Fedora 40.

Could we just target this for Fedora 41?

**Comment by rahulsundaram** [comment]
> > That's fair. Do we have a timeline for the 2.0.0 release for my understanding? Just wondering if it is worth doing a downstream patch for Fedora 40 meanwhile or just let the upstream release roll through whenever.
> 
> I don't see a world that something is safer in Fedora than it is in the upstream release. I think the same risk applies. Given Fedora 40 is around the corner, I don't think we want to potentially break Fedora 40.
> 
> Could we just target this for Fedora 41?

Yes, the same risk applies of course but the logistics of how we are collecting feedback is different depending on the approach.  It's absolutely fine to wait till Fedora 41.

**Comment by hughsie** [comment]
FWIW, I'm planning on fwupd 1.9.x for F39,F40 and RHEL 9.X, and 2.0.x for F41.

**Pull request merged**